File: backend/src/main/java/com/coing/infra/upbit/handler/UpbitWebSocketTickerHandler.java
package com.coing.infra.upbit.handler;

import java.nio.charset.StandardCharsets;

import org.springframework.stereotype.Component;
import org.springframework.web.socket.BinaryMessage;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;
import org.springframework.web.socket.handler.BinaryWebSocketHandler;

import com.coing.infra.upbit.adapter.UpbitDataService;
import com.coing.infra.upbit.dto.UpbitWebSocketTickerDto;
import com.coing.infra.upbit.enums.EnumUpbitRequestType;
import com.coing.infra.upbit.util.UpbitRequestBuilder;
import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Component
@RequiredArgsConstructor
@Slf4j
public class UpbitWebSocketTickerHandler extends BinaryWebSocketHandler {
	private final ObjectMapper objectMapper;
	private final UpbitDataService upbitDataService;
	private final UpbitRequestBuilder upbitRequestBuilder;

	@Override
	public void afterConnectionEstablished(WebSocketSession session) throws Exception {
		log.info("Upbit WebSocket Ticker connection established.");
		String subscribeMessage = upbitRequestBuilder.makeRequest(EnumUpbitRequestType.TICKER);
		session.sendMessage(new TextMessage(subscribeMessage));
	}

	@Override
	public void handleBinaryMessage(WebSocketSession session, BinaryMessage message) {
		String payload = new String(message.getPayload().array(), StandardCharsets.UTF_8);

		if (!payload.isEmpty()) {
			processMessage(payload);
		}
	}

	private void processMessage(String payload) {
		try {
			if ("{\"status\":\"UP\"}".equals(payload)) {
				log.debug("Received keepalive message: {}", payload);
				return;
			}

			UpbitWebSocketTickerDto tickerDto = objectMapper.readValue(payload, UpbitWebSocketTickerDto.class);
			upbitDataService.processTickerData(tickerDto);
		} catch (Exception e) {
			log.error("Error processing ticker message: {}", payload, e);
		}
	}
}

File: backend/src/main/java/com/coing/infra/upbit/handler/UpbitWebSocketTickerHandler.kt
package com.coing.infra.upbit.handler

import com.coing.infra.upbit.adapter.UpbitDataService
import com.coing.infra.upbit.dto.UpbitWebSocketTickerDto
import com.coing.infra.upbit.enums.EnumUpbitRequestType
import com.coing.infra.upbit.util.UpbitRequestBuilder
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.socket.BinaryMessage
import org.springframework.web.socket.TextMessage
import org.springframework.web.socket.WebSocketSession
import org.springframework.web.socket.handler.BinaryWebSocketHandler
import java.nio.charset.StandardCharsets

@Component
class UpbitWebSocketTickerHandler(
    private val objectMapper: ObjectMapper,
    private val upbitDataService: UpbitDataService,
    private val upbitRequestBuilder: UpbitRequestBuilder
) : BinaryWebSocketHandler() {

    private val log = LoggerFactory.getLogger(this::class.java)

    override fun afterConnectionEstablished(session: WebSocketSession) {
        log.info("Upbit WebSocket Ticker connection established.")
        val subscribeMessage = upbitRequestBuilder.makeRequest(EnumUpbitRequestType.TICKER)
        session.sendMessage(TextMessage(subscribeMessage))
    }

    override fun handleBinaryMessage(session: WebSocketSession, message: BinaryMessage) {
        val payload = String(message.payload.array(), StandardCharsets.UTF_8)
        if (payload.isNotEmpty()) processMessage(payload)
    }

    private fun processMessage(payload: String) {
        try {
            if (payload == """{"status":"UP"}""") {
                log.debug("Received keepalive message: {}", payload)
                return
            }

            val tickerDto = objectMapper.readValue(payload, UpbitWebSocketTickerDto::class.java)
            upbitDataService.processTickerData(tickerDto)
        } catch (e: Exception) {
            log.error("Error processing ticker message: {}", payload, e)
        }
    }
}
